kv: fix local server optimization with --advertise-addr

tbg opened the pull request:
We were using `Addr` to match against our local addr, but should have been using AdvertiseAddr.

Fixes #19991.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20006)
<!-- Reviewable:end -->

a-robinson commented on pkg/rpc/context_test.go:
```diff
@@ -110,6 +110,38 @@ func (*internalServer) Batch(
 }
 
+// TestInternalServerAddress verifies that RPCContext uses AdvertiseAddr, not Addr, to
+// determine whether the apply the local server optimization.
```
s/the apply/to apply/ if you have to burn another CI cycle

a-robinson commented on pkg/rpc/context_test.go:
```diff
@@ -110,6 +110,38 @@ func (*internalServer) Batch(
+	internal := &internalServer{}
+	serverCtx.SetLocalInternalServer(internal)
+
+	if is := serverCtx.GetLocalInternalServerForAddr(baseCtx.Addr); is != nil {
```
It strikes me that this test case is kind of the opposite of what we want. We don't actually expect `baseCtx.Addr` to be the target of any traffic, but if it was I don't know why we wouldn't want to use the local server optimization for it.

More of an observation than anything, though, given that we don't expect it to be the target of real traffic.

a-robinson reviewed (approved):
LGTM

nvanbenschoten commented on pkg/rpc/context_test.go:
```diff
@@ -215,7 +248,10 @@ func TestHeartbeatHealth(t *testing.T) {
 
-	if err := clientCtx.ConnHealth(clientCtx.Addr); err != nil {
+	if err := clientCtx.ConnHealth(clientCtx.Addr); err != ErrNotConnected {
+		t.Error(err)
```
Switch this to a `t.Errorf` and explain that the error should have been `ErrNotConnected`, otherwise, if it returns no error, the test failure will be very unclear.

nvb reviewed (approved):
Nice find!

tbg committed:
kv: fix local server optimization with --advertise-host

We were using `Addr` to match against our local addr, but should have been using AdvertiseAddr.

Release Note (bug fix): enable an RPC-saving optimization even when `--advertise-host` is used.

Fixes #19991.

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/rpc/context_test.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20006#-Kyq4kDc2ainQ6SpSpxe-r1-114:-KyqOmWHi90z8zVt398D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb1d6ab599085045ba6d481e581fffe0d1ecae6d/pkg/rpc/context_test.go#L114)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/the apply/to apply/ if you have to burn another CI cycle
</blockquote></details>

Done.

---

*[pkg/rpc/context_test.go, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20006#-Kyq4kDc2ainQ6SpSpxe-r1-136:-KyqOq3Dc3OAl3Wc9y6T:bfcqqov) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb1d6ab599085045ba6d481e581fffe0d1ecae6d/pkg/rpc/context_test.go#L136)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It strikes me that this test case is kind of the opposite of what we want. We don't actually expect `baseCtx.Addr` to be the target of any traffic, but if it was I don't know why we wouldn't want to use the local server optimization for it.

More of an observation than anything, though, given that we don't expect it to be the target of real traffic.
</blockquote></details>

Agreed, removed.

---

*[pkg/rpc/context_test.go, line 252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20006#-Kyq4kDc2ainQ6SpSpxe-r1-252:-KyqbPvNGsyhihhasNjr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb1d6ab599085045ba6d481e581fffe0d1ecae6d/pkg/rpc/context_test.go#L252)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Switch this to a `t.Errorf` and explain that the error should have been `ErrNotConnected`, otherwise, if it returns no error, the test failure will be very unclear.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20006)*
<!-- Sent from Reviewable.io -->
